### ISSUE TITLE
Feature: handle .prettierignore location like .gitignore and .npmignore

### ISSUE THREAD

**Issue by bebbi** [prettier/prettier#4081]
Suggested feature: handle .prettierignore files like .gitignore and .npmignore files, allowing them at any level of the project, not only root.

Use-case:
Using prettier in sub-projects and their "container" projects, where they are editable both from within a sub-projects repository, as well as the container repository.
Examples: git submodules, git subtree, git subrepo, and mechanisms to sync mono- and sub-repos such as [splitsh](https://github.com/splitsh/lite)

In this case, we may need a .prettierignore file that is specific to a sub-project and is located at the sub-tree root rather than the project root. Otherwise, we need duplicate .prettierignore declarations with different paths in the sub-project repo and in the container repo.

From [git docs](https://git-scm.com/docs/gitignore):

```
Patterns read from a .gitignore file in the same directory as the path, or in any parent directory,
with patterns in the higher level files (up to the toplevel of the work tree) being overridden
by those in lower level files down to the directory containing the file.
These patterns match relative to the location of the .gitignore file.
A project normally includes such .gitignore files in its repository,
containing patterns for files generated as part of the project build.
```

From [npm docs](https://docs.npmjs.com/misc/developers):
```
Like git, npm looks for .npmignore and .gitignore files in all subdirectories of your package,
not only the root directory.
```

**Comment by lipis** [comment]
Great idea..

**Comment by ikatyang** [comment]
This looks like a duplicate of #3274.

**Comment by duailibe** [comment]
It's not exactly a duplicate since #3274 is more of a question than a feature request..

I guess it's worth discussing if we should implement this

**Comment by lipis** [comment]
Especially useful in monorepos.

**Comment by duailibe** [comment]
Usually with monorepos you have a single `.prettierignore` and Prettier is installed in the root of the packages. So it's not really useful in those cases.

The use case the OP posted though (sub-repos may be edited "isolated") it might be useful but it's more rare.

I'm not convinced we need this feature but I don't oppose.

**Comment by bebbi** [comment]
FWIW, we use git subtrees and similar in monorepos when externals collaborate on isolated sub-projects.

**Comment by ericanderson** [comment]
Hmm, an extension of this might be to find the .prettierrc per file location. Thus:
```
child/
  src/
    file.js
  .prettierrc
src/
  file.js
.prettierrc
```

**Comment by lipis** [comment]
@ericanderson  I think that one works.. (not sure though)

**Comment by suchipi** [comment]
This feature would be nice to have, mostly for monorepos, but also for deeply nested build output directories. My first intuition was that this would be supported (because I was used to .gitignore), but it was not.

**Comment by fahrradflucht** [comment]
The question that needs to be discussed before it is implementable is how this should interact with the `--ignore-path`cli flag.

**Comment by bebbi** [comment]
Is this feature still under consideration?

**Comment by suchipi** [comment]
Yes

**Comment by bmalehorn** [comment]
@bebbi, @suchipi I've wanted to fix this issue for a long time, so I took a go at fixing it: https://github.com/prettier/prettier/pull/6203

`.prettierignore` works just like I expected now, pretty much like `.gitignore`!

The one edge case is `--ignore-path`. I decided that if `--ignore-path` is specified, the "`.prettierignore` at any level" feature is disabled and we only use the given file. This gives users an escape hatch back to the old behavior by specifying `--ignore-path .prettierignore`.

Thoughts? What kind of behavior would you expect?

**Comment by lucasconstantino** [comment]
By the way: until @bmalehorn work isn't merged, I'm currently adding a symbolic link to from the `./root/.prettierignore` to `./root/sub-project/.prettierignore` for this to work... quite ugly, but works.

**Comment by tinynumbers** [comment]
Any updates on this PR, or plans to merge it?

**Comment by ggabriel96** [comment]
Multi-language projects would benefit from this too.

I am struggling to configure linting for both Python and Node.js, since the working directory interfere with the behaviour of the tools.
For example, I can control the working directory in a CI job, but not in a local pre-commit configuration. Everything works fine if all tools search the project directories for their configuration files.

**Comment by ArnaudVerpouckeDeveloper** [comment]
Could it be, that this is still not fixed?

**Comment by codingwithchris** [comment]
Is this still under consideration? With tools like Turborepo getting more and more popular, it's very common to have a prettier ignore in each sub-package so prettier can be run in parallel with different configs and ignores.

In our case, we have the setup described above we also have a prettier ignore in the root of our monorepo since we run prettier at the root with lint-staged and husky to catch root formatting issues. BUT... setting it up in this way is causing prettier to only ever use our root ignore and totally forget about our other sub-package prettier ignores. Very frustrating.

It seems to detect sub-packages configs just fine, however -- so it's very odd that the ignore file doesn't behave in the same way.

**Comment by TomSssM** [comment]
Since prettier seems to resolve its config file up the directory tree I found this tricky solution to the problem, just write to `.prettierrc` / `.prettierrc.js` etc.
```jsonc
{
  "overrides": [
    {
      "files": ["*.js", "*.json"], // all the files you want
      "options": {
          // options go here...
      },
      "excludeFiles": ["/dist"] // analog to .prettierignore
    }
  ]
}
```
Since it is not easy to remember better to put this setup into shareable prettier config

**Comment by elisei-timofte** [comment]
Any updates on this one?

**Comment by system32uwu** [comment]
Ugly but works for monorepos (I'm using turborepo):

```json
"scripts": {
    "format": "prettier --ignore-path <(cat ../../.prettierignore) --write ."
}
```

Make sure to get the path right, in my case the structure is the following

```
|_ .prettierignore
   \_apps
       \_landing
```

[Credits](https://github.com/prettier/prettier/issues/8048#issuecomment-1120339540)

**Comment by sscotth** [comment]
Nice workaround. If only that worked with vscode format on save.

**Comment by wottpal** [comment]
Another workaround when working with VSCode is working inside `/.vscode/XXX.code-workspace`-files instead. In this environment, VSCode & prettier behave as they should! 

Btw this also has many more benefits (e.g. it splits the tree-view into different packages, you can quickly open new terminal sessions on specific package-levels etc.). 

More info: https://code.visualstudio.com/docs/editor/workspaces

**Comment by httpete** [comment]
I just plop a .prettierignore at every level. It's flexible.

**Comment by AlexAegis** [comment]
This would likely solve the problem when in a monorepo you run prettier in every package but you only have a config and ignore file at the repo root. (this is really apparent in turborepo managed monorepos) 

Consider this scenario:

```sh
<reporoot>/.prettierrc.cjs
<reporoot>/.prettierignore
<reporoot>/packages/foo (cwd when turbo runs a task, this is where stuff is formatted & checked)
```

Prettier does pick up the config file two levels above but not the ignore file.

For now I'm forced to copy the ignore file in every package folder

**Comment by m3hari** [comment]
It's bummer how inconsistent this behavior is; it resolves the config file recursively upwards but not the ignore file

**Comment by jmiller-nexient** [comment]
Up-voting getting this resolved finally.

**Comment by andrewdbond** [comment]
Workaround:

1. Install `find-up-cli` (<https://www.npmjs.com/package/find-up-cli>)
2. Use the prettier argument `--ignore-path $(find-up .prettierignore)`

To enable this in npm scripts on Windows, run `npm config set script-shell pwsh` (uses PowerShell [instead of cmd](https://docs.npmjs.com/cli/v9/using-npm/config#script-shell) to allow the substitution).

find-up-cli, by the prolific @sindresorhus, finds a file by walking up parent directories.

**Comment by matthewbordas** [comment]
Any updates on this? This is a very perplexing issue to still be running into 5+ years after it was reported and this behavior is common for other tools.

**Comment by andrewdbond** [comment]
@matthewbordas does the workaround in https://github.com/prettier/prettier/issues/4081#issuecomment-1502020203 help?

**Comment by bogacg** [comment]
@andrewdbond adding another package **and** making a config change would not be considered a workaround for many I suppose. I copy paste in monorepo packages when I setup thinking this is a senseless situation to exist in the first place.

**Comment by sscotth** [comment]
@andrewdbond @bogacg It also doesn't handle auto-formatting by editors like vscode

**Comment by niklasholm** [comment]
> Workaround:
> 
> 1. Install `find-up-cli` (<https://www.npmjs.com/package/find-up-cli>)
> 2. Use the prettier argument `--ignore-path $(find-up .prettierignore)`
> 
> To enable this in npm scripts on Windows, run `npm config set script-shell pwsh` (uses PowerShell [instead of cmd](https://docs.npmjs.com/cli/v9/using-npm/config#script-shell) to allow the substitution).
> 
> find-up-cli, by the prolific @sindresorhus, finds a file by walking up parent directories.

If your `.prettierignore` contains paths, which are relative, then this doesn't work. Prettier doesn't consider the location of the ignore file.

**Comment by xav-ie** [comment]
Can this be closed with specific reasons given to _not_ implement this feature?

I am frustrated by the false hope :) and want some resolution.

Based on this comment, https://github.com/prettier/prettier/pull/6203#issuecomment-1036586912, it looks like some mantainers did not want this feature. If a reasonable explanation can be given why not, that will help me become less frustrated, and I can be happy with workarounds. 

The main concern I can see is for people who already have repos with sub `.prettierignore` files. Valid, yes, but I also think that the current behavior does not align with standard expectations. I would expect, like many other `.*ignore` formats, it is recursive. This expectation is furthered because `.prettierignore` already says it adheres to `.gitignore` syntax. It is also muddied by the fact the some files seem to follow the recursive expectation:
https://github.com/prettier/prettier/issues/4081#issuecomment-1384632159

Maybe there can be a quip about, "This won't work recursively like `.gitignore`. It only works at top-level" on https://prettier.io/docs/en/ignore.html? I think this can help a lot if that is the direction wanted.

I am just confused because this seems to be in a weird limbo and don't understand why. Thank you :)

**Comment by foxysolutions** [comment]
Just figured that while `.prettierignore` files were not picked up in my project with multiple nested folders, there was a way to 'exclude' one specific folder. This is a Salesforce project, which has code put away in the `force-app/main/default` folder.

The way to exclude all files except code files, this can be realised via:
```
# Ignore full repository, except for the force-app folder
/*
!/force-app

# Then some wacky .prettierignore logic, as is it not recursive, so we should 'guide' the ignore logic towards it
/force-app/*
!/force-app/main
/force-app/main/default/*

# Finally 'exclude' the code-folders from being ignored, thus apply prettier onto them
!/force-app/main/default/aura
!/force-app/main/default/classes
...
```

Might not be fully related to this topic, but might help some Googling persons to find a workaround to this is resolved/supported.

**Comment by slifty** [comment]
I just wanted to share another use case here -- I have a node project but we're using a python dependency (sqlfluff), and our project is intentionally un-opinionated about how to run that dependency.  In my case I'm using `venv` which means I have a `/venv` directory.

Prettier is processing and finding issues with some of the generated files in `venv`

Since `venv` is a local artifact that is created because of my personal tooling choices, I want it to be ignored AND I do not want the code base to have an explicit `ignore` for it as I am the only person who faces it.  I believe is the responsibility of individual developers to configure their local tooling in a way that does not result in commits to accommodate their tools.  For example, venv is not in our project `.gitignore` but is instead in my global `~/.gitignore`.

For other tools I'm able to add a `.fooIgnore` to `/venv` which means I can tell the tools to ignore that directory.

It seems prettier does not have any way for me to accomplish this without modifying my upstream repository.

Providing this flexibility, or at least allowing for a global `.prettierignore` would allow individual contributors to manage their own developer tools.

**Comment by kfogel** [comment]
> I just wanted to share another use case here -- I have a node project but we're using a python dependency (sqlfluff), and our project is intentionally un-opinionated about how to run that dependency. In my case I'm using `venv` which means I have a `/venv` directory.
> 
> Prettier is processing and finding issues with some of the generated files in `venv`
> 
> Since `venv` is a local artifact that is created because of my personal tooling choices, I want it to be ignored AND I do not want the code base to have an explicit `ignore` for it as I am the only person who faces it. I believe is the responsibility of individual developers to configure their local tooling in a way that does not result in commits to accommodate their tools. For example, venv is not in our project `.gitignore` but is instead in my global `~/.gitignore`.
> 
> For other tools I'm able to add a `.fooIgnore` to `venv/` which means I can tell the tools to ignore that directory.
> 
> It seems prettier does not have any way for me to accomplish this without modifying my upstream repository to accommodate my personal tooling.
> 
> Providing the flexibility of subdirectory ignore files, or at least allowing for a global `.prettierignore` would allow individual contributors to manage their own developer tools.

Or... an option to have prettier parse & obey `.gitignore` files?  Probably for most developers, whatever Git is ignoring they'd want prettier to ignore too.

**Comment by Cryszon** [comment]
> Or... an option to have prettier parse & obey one's `~/.gitignore`? Probably for most developers, whatever Git is ignoring they'd want prettier to ignore too.

For most developers, probably. But simply using `.gitignore` is not enough when you want Prettier to ignore something tracked by Git ([e.g. lock files for alternative package managers](https://github.com/prettier/prettier/issues/14517)).

**Comment by Leksat** [comment]
I tried creating an external workaround for this issue and discovered that https://www.npmjs.com/package/ignore-walk can list all non-ignored files in a monorepo. With a single line of code:

```ts
import ignoreWalk from 'ignore-walk';

const paths = ignoreWalk.sync({ path: '.', ignoreFiles: ['.gitignore', '.prettierignore'] })
```

This respects all nested ignore files ✅

I'm not sure how Prettier currently discovers files to parse, but if it can leverage `ignore-walk`, wouldn't that be an easy win?

**Comment by Agritite** [comment]
> I just wanted to share another use case here -- I have a node project but we're using a python dependency (sqlfluff), and our project is intentionally un-opinionated about how to run that dependency. In my case I'm using `venv` which means I have a `/venv` directory.
> 
> Prettier is processing and finding issues with some of the generated files in `venv`
> 
> Since `venv` is a local artifact that is created because of my personal tooling choices, I want it to be ignored AND I do not want the code base to have an explicit `ignore` for it as I am the only person who faces it. I believe is the responsibility of individual developers to configure their local tooling in a way that does not result in commits to accommodate their tools. For example, venv is not in our project `.gitignore` but is instead in my global `~/.gitignore`.
> 
> For other tools I'm able to add a `.fooIgnore` to `venv/` which means I can tell the tools to ignore that directory.
> 
> It seems prettier does not have any way for me to accomplish this without modifying my upstream repository to accommodate my personal tooling.
> 
> Providing the flexibility of subdirectory ignore files, or at least allowing for a global `.prettierignore` would allow individual contributors to manage their own developer tools.

I use `uv` and it creates `.venv` and `uv.lock`. For git I can add those to `.git/info/exclude` to not affect upstream, but for prettier there's currently no way to ignore those files besides stashing `.prettierignore ` everytime I want to commit.

**Comment by danguilherme** [comment]
Please, can we have this? 🥹

**Comment by alvarosevilla95** [comment]
+1 on requesting this. in my case, this would be particularly useful when using a combination of lint-staged and turborepo.

with turbo, the prettier command is configured to to run from each package in a monorepo, so it will ignore a root prettierignore and only use the one configured in the subpackage.

with lint-staged, it's easier to run from the root of the repo, rather than running lint-staged on every subpackage (reduces the number of processes started), so that will use the root prettierignore.

as such, currently i need to have all my prettierignore settings duplicated at root and the package level. if prettier was able to lookup the appropriate file instead (like [eslint](https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution) now supports), it would simplify its use for monorepos

**Comment by bestickley** [comment]
Idea: since prettier recursively looks up directories for prettier configuration file (documented [here](https://prettier.io/docs/configuration)), could `ignore` property be added to prettier configuration file (i.e. .prettierrc.js) to achieve this feature? 

The ability to define global prettier ignore configuration for a workspace reduces a lot of duplicate configuration in monorepos with turborepo (or nx) where you want to cache formatting per package and not have to run prettier from root workspace level each time as recommended in the [related PR](https://github.com/prettier/prettier/pull/6203#issuecomment-1036586912).

**Comment by Manikhan1322** [comment]
> Suggested feature: handle .prettierignore files like .gitignore and .npmignore files, allowing them at any level of the project, not only root.
> 
> Use-case: Using prettier in sub-projects and their "container" projects, where they are editable both from within a sub-projects repository, as well as the container repository. Examples: git submodules, git subtree, git subrepo, and mechanisms to sync mono- and sub-repos such as [splitsh](https://github.com/splitsh/lite)
> 
> In this case, we may need a .prettierignore file that is specific to a sub-project and is located at the sub-tree root rather than the project root. Otherwise, we need duplicate .prettierignore declarations with different paths in the sub-project repo and in the container repo.
> 
> From [git docs](https://git-scm.com/docs/gitignore):
> 
> ```
> Patterns read from a .gitignore file in the same directory as the path, or in any parent directory,
> with patterns in the higher level files (up to the toplevel of the work tree) being overridden
> by those in lower level files down to the directory containing the file.
> These patterns match relative to the location of the .gitignore file.
> A project normally includes such .gitignore files in its repository,
> containing patterns for files generated as part of the project build.
> ```
> 
> From [npm docs](https://docs.npmjs.com/misc/developers):
> 
> ```
> Like git, npm looks for .npmignore and .gitignore files in all subdirectories of your package,
> not only the root directory.
> ```

**Comment by Leksat** [comment]
> I tried creating an external workaround for this issue

And I ended up creating one 😁

https://www.npmjs.com/package/prettierignore-monorepo

Tested it on several projects. Works for me. Hopefully it can be useful to others 🤞

**Comment by victorandree** [comment]
I'm wondering if it would be possible to somehow hack this functionality using a plugin?

A sketch of an idea would be to provide a custom `printer` where you should have access to `options.filepath` in `print(path, options, print) {}` and it might be possible to just... print the thing as-is.

**Comment by mmaryo** [comment]
good idea

**Comment by The-Best-Codes** [comment]
+1, it would be awesome if Prettier supported this! My project has a `.gitignore` in the project root and one in `src-tauri/`, I expected Prettier to respect both, just like git, but it formats all the files in `src-tauri/` including the gitignored ones.

**Comment by dmotte** [comment]
Just in case, I wrote a small Bash script as a workaround for this issue. It works by combining all the ignore files into a single, temporary one.

https://gist.github.com/dmotte/9af352629e1a6142810ad9dbf4e88006

For my use case (running Prettier with `npx` on Debian/Ubuntu) it works fine, but of course your use case may be different.

Also, any feedback is welcome :)
